### PR TITLE
OneHotEncoder Hypergrid Adapter integration in Regression-Enhanced Random Forest Model

### PR DESCRIPTION
**Pull request by edcthayer**
Goal here was to replace the existing one hot encoder developed in the Regression-Enhanced Random Forest (RERF) RegressionModel with the CategoricalToOneHotEncoderHypergridAdapter.   In general, this code is simpler than the previous RERF code.  

Notes on changes:
- Recall the existing RERF one hot encoder tried to infer which dimensions were categorical dimensions and this proved impossible when too few points are passed to the fit() or predict() methods.  This led to flaky test failures.  The changes in this PR should correct those flaky test events.
- Because the OneHotEncoder Adapter understands the full adaptee hypergrid structure, several "book keeping" elements in the RERF class are no longer needed.  
- The RERF unit test for gradient was removed and will be re-introduced when the path forward for using the RERF base model gradient.
- The RERF unit tests revealed that the OneHotEncoder Adapter failed when there were no categorical dimensions, so this PR remedies that defect.

### CODE REVIEW

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/RegressionEnhancedRandomForestModel.py [hunk 1]
```diff
@@ -427,7 +391,7 @@ def predict(self, feature_values_pandas_frame, include_only_valid_rows=True):
                                               set(feature_values_pandas_frame.columns.values))
         for missing_column_name in missing_column_names:
             feature_values_pandas_frame[missing_column_name] = np.NaN
-        x_df = feature_values_pandas_frame[self.input_dimension_names]
+        x_df = self.one_hot_encoder_adapter.project_dataframe(df=feature_values_pandas_frame, in_place=False)
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/RegressionEnhancedRandomForestModel.py:394]: Looks good, I created a new issue to make False a default here. #150

**Review by byte-sculptor — APPROVED** [review]

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestRegressionEnhancedRandomForestModel.py [hunk 2]
```diff
@@ -369,4 +324,4 @@ def test_lasso_hierarchical_categorical_predictions(self):
         residual_sum_of_squares = ((y_test - predicted_y) ** 2).sum()
         total_sum_of_squares = ((y_test - y_test.mean()) ** 2).sum()
         unexplained_variance = residual_sum_of_squares / total_sum_of_squares
-        self.assertTrue(unexplained_variance < 10**-4, '1 - R^2 larger than expected')
+        self.assertTrue(unexplained_variance < 10**-3, '1 - R^2 larger than expected')
```
Comment (amueller) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestRegressionEnhancedRandomForestModel.py:327]: this still seems to be failing, right?

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestRegressionEnhancedRandomForestModel.py [hunk 3]
```diff
@@ -251,12 +251,13 @@ def test_lasso_categorical_predictions(self):
...
-        num_categorical_levels_expected = len(x_train_df['x0'].unique()) * len(x_train_df['i0'].unique())
+        num_categorical_levels_expected = len(rerf.one_hot_encoder_adapter.get_one_hot_encoded_column_names())
         num_continuous_dimensions = 2  # x1 and x2
-        final_num_features = num_categorical_levels_expected - 1 + num_continuous_dimensions
+        final_num_features = num_categorical_levels_expected + num_continuous_dimensions
         polynomial_degree = self.model_config.max_basis_function_degree
         num_terms_in_polynomial_per_categorical_level = self.n_choose_k(polynomial_degree + num_continuous_dimensions, num_continuous_dimensions)
-        num_terms_in_polynomial = num_terms_in_polynomial_per_categorical_level * num_categorical_levels_expected
+        num_terms_in_polynomial = num_terms_in_polynomial_per_categorical_level * (num_categorical_levels_expected + 1)\
+                                  - len(rerf.categorical_zero_cols_idx_to_delete_)
```
Comment (amueller) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestRegressionEnhancedRandomForestModel.py:260]: Can you maybe explain this in a comment?
Is that for polynomial interactions between different levels (which will always be zero)?

**Review by amueller — COMMENTED** [review]
looks good overall but I'm confused why the tolerance needed to be increased and why it's still failing.

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestRegressionEnhancedRandomForestModel.py (hunk 2 above)
Comment (edcthayer) [reply source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestRegressionEnhancedRandomForestModel.py]: The failures were arising from ill-conditioned matrix passed to lasso regressor.  Added sklearn StandardScaler and the degree of flakiness appears to have been resolved.  I will create a hypergrid adapter (wrapper) for StandardScaler in the next PR.

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestRegressionEnhancedRandomForestModel.py (hunk 3 above)
Comment (edcthayer) [reply source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestRegressionEnhancedRandomForestModel.py:247]: Sorry.  That var name is poorly chosen.  Chose a more fitting name and added some comments.

**Pull request merged**